src/vdaf/mastic.rs

```diff
@@ -0,0 +1,727 @@
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn shard_with_random(
+        &self,
+        measurement_label: &VidpfInput,
+        measurement_weight: &VidpfWeight<T::Field>,
+        nonce: &[u8; 16],
+        vidpf_keys: [VidpfKey; 2],
+        szk_random: [Seed<SEED_SIZE>; 2],
+        opt_random: Option<Seed<SEED_SIZE>>,
+    ) -> Result<(<Self as Vdaf>::PublicShare, Vec<<Self as Vdaf>::InputShare>), VdafError> {
+        // Compute the measurement shares for each aggregator by generating VIDPF
+        // keys for the measurement and evaluating each of them.
+        let public_share = self.vidpf().gen_with_keys(
+            &vidpf_keys,
+            measurement_label,
+            measurement_weight,
+            nonce,
+        )?;
+
+        let leader_measurement_share = self
+            .vidpf()
+            .eval(
+                &vidpf_keys[0],
+                &public_share,
+                &VidpfInput::from_bools(&[false]),
+                nonce,
+            )?
+            .share
+            + self
+                .vidpf()
+                .eval(
+                    &vidpf_keys[0],
+                    &public_share,
+                    &VidpfInput::from_bools(&[true]),
+                    nonce,
+                )?
+                .share;
+        let helper_measurement_share = self
+            .vidpf()
+            .eval(
+                &vidpf_keys[1],
+                &public_share,
+                &VidpfInput::from_bools(&[false]),
+                nonce,
+            )?
+            .share
+            + self
+                .vidpf()
+                .eval(
+                    &vidpf_keys[1],
+                    &public_share,
+                    &VidpfInput::from_bools(&[true]),
+                    nonce,
+                )?
+                .share;
+
+        let szk_proof_shares = self.szk.prove(
+            leader_measurement_share.as_slice(),
+            helper_measurement_share.as_slice(),
+            measurement_weight.as_slice(),
+            szk_random,
+            opt_random,
+            nonce,
+        )?;
+        let leader_share = MasticInputShare::<T::Field, SEED_SIZE> {
+            vidpf_key: vidpf_keys[0].clone(),
+            proofs_share: szk_proof_shares[0].clone(),
+        };
+        let helper_share = MasticInputShare::<T::Field, SEED_SIZE> {
+            vidpf_key: vidpf_keys[1].clone(),
+            proofs_share: szk_proof_shares[1].clone(),
+        };
+        Ok((public_share, vec![leader_share, helper_share]))
+    }
+
+    pub(crate) fn is_randomized(&self) -> bool {
+        self.szk.has_joint_rand()
+    }
+
+    pub(crate) fn proof_len(&self) -> usize {
+        self.szk.proof_len()
+    }
+
+    fn encode_measurement(
+        &self,
+        measurement: &T::Measurement,
+    ) -> Result<VidpfWeight<T::Field>, VdafError> {
+        Ok(VidpfWeight::<T::Field>::from(
+            self.szk.typ().encode_measurement(measurement)?,
+        ))
+    }
```

Comment:
As mentioned above, I think these indirections are probably not needed, and they hurt readability somewhat.